classes/form/guestlink_access_form.php

```diff
@@ -0,0 +1,104 @@
+        $mform->setType('guestlinkurl', PARAM_TEXT);
+        $mform->setDefault('guestlinkurl', $this->_customdata['guestlink']['url']);
+        $inviteparticipantsgroup = [];
+        $inviteparticipantsgroup[] =& $inviteparticipantselement;
+        $inviteparticipantsgroup[] =& $mform->createElement('button', 'guestlinkurl_copy', get_string('mod_guestlink_access_form_copy', 'bigbluebuttonbn'));
+        $mform->addGroup($inviteparticipantsgroup, 'inviteparticipantsgroup', get_string('view_guestlink_label', 'bigbluebuttonbn'), array(' '), false);
+
+        // Access Code.
+        $accesscodeelement = $mform->createElement('text', 'password', '', [
+            'readonly' => true,
+            'size' => 29,
+            'placeholder' => get_string('view_guestlink_password_no_password_set', 'bigbluebuttonbn')
+        ]);
+        $mform->setType('password', PARAM_TEXT);
+        $mform->setDefault('password', $this->_customdata['guestlink']['password'] ?? '');
+        $accesscodegroup = [];
+        $accesscodegroup[] =& $accesscodeelement;
+        if (!empty($this->_customdata['guestlink']['changepassenabled'])) {
+            $accesscodegroup[] =& $mform->createElement('button', 'password_clear', get_string('mod_guestlink_access_form_clear', 'bigbluebuttonbn'));
+        }
+        if (!empty($this->_customdata['guestlink']['changepassenabled'])) {
+            $accesscodegroup[] =& $mform->createElement('button', 'password_change', get_string('mod_guestlink_access_form_generate', 'bigbluebuttonbn'));
+        }
+        $accesscodegroup[] =& $mform->createElement('button', 'password_copy', get_string('mod_guestlink_access_form_copy', 'bigbluebuttonbn'));
+        $mform->addGroup($accesscodegroup, 'accesscodegroup', get_string('view_guestlink_password_label', 'bigbluebuttonbn'), array(' '), false);
+
+        // Expires At.
+        $defaultdateoptions = [
+           'optional' => true
+        ];
+        $expiresatelement = $mform->createElement('date_time_selector', 'guestlinkexpiresat', '', $defaultdateoptions);
+        $mform->setType('guestlinkexpiresat', PARAM_RAW);
+        $accesscodegroup = [];
+        $accesscodegroup[] =& $expiresatelement;
+        $mform->addGroup($accesscodegroup, 'accesscodegroup', get_string('view_guestlink_expires_at_label', 'bigbluebuttonbn'), array(' '), false);
+
+        $this->add_action_buttons($cancel = false, get_string('view_guestlink_save_settings', 'bigbluebuttonbn'));
+
+        // CM ID.
+        $mform->addElement('hidden', 'id');
+        $mform->setType('id', PARAM_ALPHANUMEXT);
+        $mform->setDefault('id', $this->_customdata['bbbsession']['cm']->id);
+
+        // BBB Instance ID.
+        $mform->addElement('hidden', 'bn');
+        $mform->setType('bn', PARAM_ALPHANUMEXT);
+        $mform->setDefault('bn', $this->_customdata['bbbsession']['bigbluebuttonbn']->id);
```

Comment:
Just a personal choice,. but I think the _customdata should just contain the actual values, not the whole bbbsession object. The form shouldnt really need to know the structure of the parent data obj, just the exact data that it cares about. I would move this array index selection into the customdata construction in viewlib